Orange/widgets/visualize/widgetutils/common/view.py

```diff
@@ -0,0 +1,180 @@
+from itertools import repeat
+
+import numpy as np
+from PyQt4 import QtGui
+from PyQt4.QtCore import Qt
+
+
+class ZoomableGraphicsView(QtGui.QGraphicsView):
+    """Zoomable graphics view.
+
+    Composable graphics view that adds zoom functionality.
+
+    It also handles automatic resizing of content whenever the window is
+    resized.
+
+    Right click will reset the zoom to a factor where the entire scene is
+    visible.
+
+    Parameters
+    ----------
+    scene : QtGui.QGraphicsScene
+    padding : int or tuple, optional
+        Specify the padding around the drawn widgets. Can be an int, or tuple,
+        the tuple can contain either 2 or 4 elements.
+
+    Notes
+    -----
+      - This view will consume wheel scrolling and right mouse click events.
+
+    """
+
+    def __init__(self, scene, padding=(0, 0), **kwargs):
+        self.zoom = 1
+        self.scale_factor = 1 / 16
+        # zoomout limit prevents the zoom factor to become negative, which
+        # results in the canvas being flipped over the x axis
+        self.__zoomout_limit_reached = False
+        # Does the view need to recalculate the initial scale factor
+        self.__needs_to_recalculate_initial = True
+        self.__initial_zoom = -1
+
+        self.__central_widget = None
+        self.__set_padding(padding)
+
+        super().__init__(scene, **kwargs)
+
+    def resizeEvent(self, ev):
+        super().resizeEvent(ev)
+        self.__needs_to_recalculate_initial = True
+
+    def wheelEvent(self, ev):
+        self.__handle_zoom(ev.delta())
+        super().wheelEvent(ev)
+
+    def mousePressEvent(self, ev):
+        # right click resets the zoom factor
+        if ev.button() == Qt.RightButton:
+            self.reset_zoom()
+        super().mousePressEvent(ev)
+
+    def keyPressEvent(self, ev):
+        if ev.key() == Qt.Key_Plus:
+            self.__handle_zoom(1)
+        elif ev.key() == Qt.Key_Minus:
+            self.__handle_zoom(-1)
+
+        super().keyPressEvent(ev)
+
+    def __set_padding(self, padding):
+        # Allow for multiple formats of padding for convenience
+        if isinstance(padding, int):
+            padding = list(repeat(padding, 4))
+        elif isinstance(padding, list) or isinstance(padding, tuple):
+            if len(padding) == 2:
+                padding = (*padding, *padding)
```

Comment:
This is (unfortunately) not Python 3.4 syntax. Accommodating people using LTS software (including myself), we should keep this Python 3.4-compatible.